src/lib/core/prbuf.c

```diff
@@ -369,3 +376,170 @@ prbuf_iterator_next(struct prbuf_iterator *iter, struct prbuf_entry *result)
 	result->ptr = iter->current->data;
 	return 0;
 }
+
+void
+prbuf_reader_close(struct prbuf_reader_iterator *iter)
+{
+	ibuf_destroy(&iter->buf);
+}
+
+/**
+ * Check whether buffer header is sane.
+ */
+static int
+prbuf_check_header(struct prbuf_header *header)
+{
+	if (header->version > prbuf_version) {
+		diag_set(IllegalParams,
+			 "unknown format version %d", header->version);
+		return -1;
+	}
+
+	/* Check we can the read first record and end position is correct. */
+	if (prbuf_header_size + header->begin +
+				record_size_overhead >= header->size ||
+	    prbuf_header_size + header->end > header->size) {
+		diag_set(IllegalParams, "inconsistent header");
+		return -1;
+	}
+
+	return 0;
+}
+
+/** Read ahead is 128k. */
+#define PRBUF_READ_AHEAD (1 << 17)
+
+int
+prbuf_reader_create(struct prbuf_reader_iterator *iter, int fd, off_t offset)
+{
+	struct prbuf_header header;
+	/*
+	 * Read ahead does not make sense when reading prbuf header as
+	 * regularly first record is not at the beginning of the buffer.
+	 */
+	ssize_t rc = fio_pread(fd, &header, prbuf_header_size, offset);
+	if (rc == -1) {
+		diag_set(SystemError, "read failed");
+		return -1;
+	}
+	if ((size_t)rc < prbuf_header_size) {
+		diag_set(IllegalParams, "can't read buffer header");
+		return -1;
+	}
+	if (prbuf_check_header(&header) == -1)
+		return -1;
+
+	ibuf_create(&iter->buf, &cord()->slabc, PRBUF_READ_AHEAD);
+	iter->data_begin = offset + prbuf_header_size;
+	iter->data_end = offset + header.size;
+	iter->pos = iter->data_begin + header.begin;
+	iter->read_pos = iter->pos;
+	iter->fd = fd;
+
+	if (header.end == 0)
+		iter->unread_size = 0;
```

Comment:
Nit: I think this case should be covered by the `header.begin >= header.end` condition.

Reply:
`header.end == 0` is a special case of empty buffer. On branch `header.begin >= header.end` `unread_size` will be calculated to full size of data area of the buffer while it should be `0`.

In particular `test_buffer_prepared_large` fails.